server-utils/src/suspendable_stream.rs

```diff
@@ -42,13 +42,17 @@ where
 
 	fn poll_next(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Option<Self::Item>> {
 		loop {
-			if let Some(timeout) = &self.timeout {
-				let timeout = tokio::time::Instant::from_std(*timeout);
-				let sleep = tokio::time::sleep_until(timeout);
+			// If we encountered a connection error before then we suspend
+			// polling from the underlying stream for a bit
+			if let Some(deadline) = &mut self.suspended_until {
+				let deadline = tokio::time::Instant::from_std(*deadline);
+				let sleep = tokio::time::sleep_until(deadline);
 				futures::pin_mut!(sleep);
 				match sleep.poll(cx) {
 					Poll::Pending => return Poll::Pending,
-					Poll::Ready(()) => {}
+					Poll::Ready(()) => {
+						self.suspended_until = None;
```

Comment:
If we're ready then let's reset the suspension flag not to execute that code path everytime